File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/ci/Coverage.java
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.gradle.plugins.ci;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.firebase.gradle.plugins.FirebaseLibraryExtension;
import java.io.File;
import java.util.List;
import org.gradle.api.Project;
import org.gradle.api.tasks.testing.Test;
import org.gradle.testing.jacoco.plugins.JacocoPluginExtension;
import org.gradle.testing.jacoco.plugins.JacocoTaskExtension;
import org.gradle.testing.jacoco.tasks.JacocoReport;

public final class Coverage {
  private Coverage() {}

  public static void apply(FirebaseLibraryExtension firebaseLibrary) {
    Project project = firebaseLibrary.project;
    project.apply(ImmutableMap.of("plugin", "jacoco"));
    File reportsDir = new File(project.getBuildDir(), "/reports/jacoco");
    JacocoPluginExtension jacoco = project.getExtensions().getByType(JacocoPluginExtension.class);

    jacoco.setToolVersion("0.8.5");
    jacoco.setReportsDir(reportsDir);
    project
        .getTasks()
        .withType(
            Test.class,
            test -> {
              JacocoTaskExtension testJacoco =
                  test.getExtensions().getByType(JacocoTaskExtension.class);
              testJacoco.setExcludeClassLoaders(ImmutableList.of("jdk.internal.*"));
              testJacoco.setIncludeNoLocationClasses(true);
            });

    project
        .getTasks()
        .create(
            "checkCoverage",
            JacocoReport.class,
            task -> {
              task.dependsOn("check");
              task.setDescription("Generates JaCoCo check coverage report.");
              task.setGroup("verification");

              List<String> excludes =
                  ImmutableList.<String>builder()
                      .add("**/R.class")
                      .add("**/R$*.class")
                      .add("**/BuildConfig.*")
                      .add("**/proto/**")
                      .add("**Manifest*.*")
                      .build();

              task.setClassDirectories(
                  project.files(
                      project.fileTree(
                          ImmutableMap.of(
                              "dir",
                              project.getBuildDir() + "/intermediates/javac/release",
                              "excludes",
                              excludes)),
                      project.fileTree(
                          ImmutableMap.of(
                              "dir",
                              project.getBuildDir() + "/tmp/kotlin-classes/release",
                              "excludes",
                              excludes))));
              task.setSourceDirectories(project.files("src/main/java", "src/main/kotlin"));
              task.setExecutionData(
                  project.fileTree(
                      ImmutableMap.of(
                          "dir",
                          project.getBuildDir(),
                          "includes",
                          ImmutableList.of("jacoco/*.exec"))));
              task.reports(
                  reports -> {
                    reports
                        .getHtml()
                        .setDestination(
                            new File(reportsDir, firebaseLibrary.artifactId.get() + "/html"));
                    reports.getXml().setEnabled(true);
                    reports
                        .getXml()
                        .setDestination(
                            new File(reportsDir, firebaseLibrary.artifactId.get() + ".xml"));
                  });
              task.getOutputs().upToDateWhen(t -> false);
            });
  }
}

File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/ci/Coverage.kt
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.gradle.plugins.ci

import com.google.firebase.gradle.plugins.FirebaseLibraryExtension
import java.io.File
import org.gradle.api.tasks.testing.Test
import org.gradle.kotlin.dsl.apply
import org.gradle.testing.jacoco.plugins.JacocoPluginExtension
import org.gradle.testing.jacoco.plugins.JacocoTaskExtension
import org.gradle.testing.jacoco.tasks.JacocoReport

fun FirebaseLibraryExtension.configureCoverage() {
    project.apply(plugin = "jacoco")
    val jacocoReportsDir = File(project.buildDir, "/reports/jacoco")

    project.extensions.configure(JacocoPluginExtension::class.java) {
        toolVersion = "0.8.5"
        reportsDir = jacocoReportsDir
    }

    project.tasks.withType(Test::class.java) {
        extensions.configure(JacocoTaskExtension::class.java) {
            excludeClassLoaders = listOf("jdk.internal.*")
            isIncludeNoLocationClasses = true
        }
    }

    project.tasks.register("checkCoverage", JacocoReport::class.java) {
        dependsOn("check")
        description = "Generates JaCoCo check coverage report."
        group = "verification"

        val excludes = listOf(
            "**/R.class",
            "**/R$*.class",
            "**/BuildConfig.*",
            "**/proto/**",
            "**Manifest*.*"
        )

        classDirectories.from(
            project.files(
                project.fileTree(
                    mapOf(
                        "dir" to project.buildDir.toString() + "/intermediates/javac/release",
                        "excludes" to excludes
                    )
                ),
                project.fileTree(
                    mapOf(
                        "dir" to project.buildDir.toString() + "/tmp/kotlin-classes/release",
                        "excludes" to excludes
                    )
                )
            )
        )
        sourceDirectories.from(project.files("src/main/java", "src/main/kotlin"))
        executionData.from(
            project.fileTree(
                mapOf(
                    "dir" to project.buildDir,
                    "includes" to listOf("jacoco/*.exec")
                )
            )
        )

        reports {
            html.destination = File(jacocoReportsDir, "${artifactId.get()}/html")
            xml.isEnabled = true
            xml.destination = File(jacocoReportsDir, "${artifactId.get()}.xml")
        }
        outputs.upToDateWhen { false }
    }
}
